src/main/java/com/conveyal/analysis/datasource/DataSourceUploadAction.java

```diff
@@ -0,0 +1,143 @@
+package com.conveyal.analysis.datasource;
+
+import com.conveyal.analysis.UserPermissions;
+import com.conveyal.analysis.models.DataSource;
+import com.conveyal.analysis.persistence.AnalysisCollection;
+import com.conveyal.file.FileStorage;
+import com.conveyal.file.FileStorageFormat;
+import com.conveyal.file.FileStorageKey;
+import com.conveyal.r5.analyst.progress.ProgressListener;
+import com.conveyal.r5.analyst.progress.TaskAction;
+import org.apache.commons.fileupload.FileItem;
+import org.apache.commons.fileupload.disk.DiskFileItem;
+import org.apache.commons.io.FilenameUtils;
+import org.bson.types.ObjectId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.conveyal.analysis.controllers.OpportunityDatasetController.getFormField;
+import static com.conveyal.analysis.datasource.SpatialLayers.detectUploadFormatAndValidate;
+import static com.conveyal.file.FileCategory.DATASOURCES;
+import static com.conveyal.file.FileStorageFormat.GEOJSON;
+import static com.conveyal.file.FileStorageFormat.SHP;
+import static com.conveyal.file.FileStorageFormat.TIFF;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Given a batch of uploaded files, put them into FileStorage, categorize and validate them, and record metadata as
+ * some specific subclass of DataSource. This implements TaskAction so it can be run in the background without blocking
+ * the HTTP request and handler thread.
+ */
+public class DataSourceUploadAction implements TaskAction {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DataSourceUploadAction.class);
+
+    // The Components used by this background task, which were formerly captured by an anonymous closure.
+    // Using named and well-defined classes for these background actions makes data flow and depdendencies clearer.
+    private final FileStorage fileStorage;
+    private final AnalysisCollection<DataSource> dataSourceCollection;
+
+    /** The files provided in the HTTP post form. These will be moved into storage. */
+    private final List<FileItem> fileItems;
+
+    /**
+     * This DataSourceIngester provides encapsulated loading and validation logic for a single format, by composition
+     * rather than subclassing. Format ingestion does not require access to the fileStorage or the database collection.
```

Comment:
No, this refers to the fact that a DataSourceIngester handling a specific format plugs into the DataSourceUploadAction, rather than having separate subclasses of DataSourceUploadAction for each format (which is how it originally worked). But then this "plugin" DataSourceIngester ended up having subclasses, so we've got both composition and inheritance here. I plan to factor a bunch of common logic out of the DataSourceIngester subclasses which may again reduce inheritance.